feat : 기본 CRUD API 개발

99winnmin opened the pull request:
## 작업사항
- ElasticSearch 를 주저장소로 하는 기본 CRUD 개발
- 프로젝트 공통 Exception 유틸기능 추가 및 핸들러 추가
- NewsService 에 대한 테스트 코드 추가

## 참고사항
- Indic 에 대해서 불변객체로 구성해서 아직 UPDATE 기능을 추가하지 못했습니다

### Review Rule

- P1: 꼭 반영해주세요 (Request changes)
- P2: 적극적으로 고려해주세요 (Request changes)
- P3: 웬만하면 반영해 주세요 (Comment)
- P4: 반영해도 좋고 넘어가도 좋습니다 (Approve)
- P5: 그냥 사소한 의견입니다 (Approve)

<!---- Resolves: #(Isuue Number) -->

99winnmin committed:
feat : elasticsearch 설정 추가

99winnmin committed:
doc : elasticsearch 의존성 추가

99winnmin committed:
doc : 기본 CRUD api-test 파일 추가

99winnmin committed:
feat : 전역 예외처리 세팅

99winnmin committed:
feat : news_articles 인덱스 엔티티 추가

99winnmin committed:
feat : 뉴스 검색,저장,삭제 end-point 추가

99winnmin committed:
feat : 뉴스 검색,저장,삭제 비지니스 로직 추가

99winnmin committed:
feat : NewsArticle 인덱스 repository 추가

99winnmin committed:
feat : 뉴스 관련 dto 집합 클래스 추가

99winnmin committed:
feat : id 와 생성,수정시간을 관리하는 BaseEntity 추가 및 불변, 비불변 추상 클래스 추가

99winnmin committed:
feat : Elasticsearch config 파일 추가

99winnmin committed:
feat : api 기본응답 data class 추가

99winnmin committed:
doc : kotlin test 의존성 추가

99winnmin committed:
test : NewsServiceTest 추가

99winnmin committed:
doc : PR 템플릿 추가

DongGeon0908 commented on src/main/kotlin/com/javacafe/realtimewinnmin/presentation/controller/NewsController.kt:
```diff
@@ -0,0 +1,120 @@
+class NewsController(
+    private val newsService: NewsService
+) {
+    private val logger = LoggerFactory.getLogger(NewsController::class.java)
```
Kotlin Logger가 더 다양한 기능을 많이 지원해욤욤

```suggestion
      private val logger = KotlinLogging.logger { }
```

p99)

99winnmin replied:
바로 차용하겠습니다!!

DongGeon0908 commented on src/main/kotlin/com/javacafe/realtimewinnmin/domain/common/BaseEntity.kt:
```diff
@@ -0,0 +1,45 @@
+/**
+ * 추가 전용 엔티티 (생성 시간만 관리)
+ */
+abstract class ImmutableEntity : BaseEntity {
```
이걸 왜 난 생각을 못했징..

DongGeon0908 replied:
👍🏻 👍🏻 👍🏻

DongGeon0908 commented on src/main/kotlin/com/javacafe/realtimewinnmin/common/dto/ApiResponse.kt:
```diff
@@ -0,0 +1,20 @@
+import java.time.LocalDateTime
+
+data class ApiResponse<T>(
+    val success: Boolean,
```
http status로는 의미를 나타내는데 한계가 있을까욤?
p5

99winnmin replied:
테스트 케이스 생각하다보니 성공,실패로만 생각하다보니 저렇게 세팅한것 같습니다..ㅎㅎ HttpStatus 로 표현하는 것이 더 좋은 것 같습니다!

DongGeon0908 commented on src/main/kotlin/com/javacafe/realtimewinnmin/common/exception/GlobalExceptionHandler.kt:
```diff
@@ -0,0 +1,21 @@
+
+    @ExceptionHandler(GlobalException::class)
+    protected fun <T> handleGlobalException(e: GlobalException): ResponseEntity<ApiResponse<T>> {
+        val response = ApiResponse.error<T>(
```
error의 interface가 모든 exception 유형마다 달라지는걸까용?
p5

99winnmin replied:
제네릭은 ApiResponse 의 data 파라미터때문이였는데 에러일 경우 data 는 필요없으므로 Nothing 으로 고정하는 식으로 처리해보겠습니다!

DongGeon0908 commented on src/main/kotlin/com/javacafe/realtimewinnmin/presentation/controller/NewsController.kt:
```diff
@@ -0,0 +1,120 @@
+                size = size
+            )
+
+            newsService.searchNews(searchRequest)
```
코드 중복이 발생할 수 있을 것 같은데, 하나의 레이어를 추가해서, 공통코드를 줄이는 방향은 어때용
구두논의
p5

99winnmin replied:
코드중복 발생을 놓쳤내요! 랩핑하는 방향으로 고려해보겠습니다!

DongGeon0908 commented on src/main/kotlin/com/javacafe/realtimewinnmin/presentation/controller/NewsController.kt:
```diff
@@ -0,0 +1,120 @@
+        @RequestParam(defaultValue = "manual") source: String = "manual",
+        @RequestParam(defaultValue = "10") size: Int = 10
+    ): ResponseEntity<ApiResponse<NewsListResponse>> {
+        logger.info("GET /api/news/search - Searching news with keyword: $keyword")
```
logger를 aop를 통해서, 녹여주는건 어떨까용? (비즈니스 코드에 더 집중하기 위함!)
p5

99winnmin replied:
controller 시작, 마지막부분은 무조건 로깅할 예정이여서 추가해보겠습니다!

99winnmin committed:
fix : adopt kotlin logger

Co-authored-by: 극락코딩 <50691225+DongGeon0908@users.noreply.github.com>

99winnmin committed:
fix : Java Logger -> Kotlin Logger 로 교체

99winnmin committed:
fix : 단순 success,fail 이 아닌 HTTP 상태 코드로 표현하도록 수정

99winnmin committed:
fix : GlobalException 처리 시 제네릭이 아닌 Nothing 으로 고정

yeonju0110 commented on src/main/kotlin/com/javacafe/realtimewinnmin/common/dto/ApiResponse.kt:
```diff
@@ -0,0 +1,36 @@
+import java.time.LocalDateTime
+
+data class ApiResponse<T>(
+    val code: HttpStatus,
```
응답 DTO에 커스텀 응답코드가 아닌 HttpStatus Code를 직접 포함하신 이유가 있을까용?

P4

99winnmin replied:
생각해보니 저는 ApiResponse 를 body 에서만 쓰는데 아예 제외하는 것도 괜찮은 것 같습니다! 감사합니다!

yeonju0110 commented on src/main/kotlin/com/javacafe/realtimewinnmin/application/service/NewsService.kt:
```diff
@@ -0,0 +1,63 @@
+
+        return NewsListResponse(
+            news = searchResults.map(NewsResponse::from),
+            totalCount = searchResults.size.toLong(),
```
여기서는 뉴스 전체 totalCount를 보내주고 있는건가요? 아니면 현재 페이지에 있는 아이템 개수를 보내주는 건가용?
totalCount라는 변수명은 전자의 의미로 쓰일 것 같아서요!

제가 elastic search는 잘 몰라서 그러는데, 지금 코드는 현재 페이지에 있는 아이템 개수를 보내주고 있는 것 같아서 궁금해서 질문 드립니당 👀

P4

99winnmin replied:
페이지네이션 생각하고 만들었는데 기능의 의미를 보아하니 단순 검색인 것 같아서 로직 수정해보겠습니다!~

yeonju0110 reviewed (approved):
코틀린의 다양한 문법 잘 봤습니다! 👍🏻

전반적으로 코드가 깔끔하고 읽기 쉽네요, 고생하셨습니다!

99winnmin committed:
fix : 불필요한 code 제거

99winnmin committed:
refactor: 뉴스 검색을 페이지네이션에서 제한된 검색으로 변경

- searchNews → searchNewsWithLimit 리네이밍
- Pageable 제거하고 size 기반 제한 적용
- 반환 타입 단순화: NewsListResponse → List<NewsResponse>
- 테스트 케이스 업데이트

da9dac reviewed (approved):
코틀린 문법을 잘 몰라서 정확하게는 모르겠지만 확실히 코드가 간결해지고 보기가 깔끔하네요!
월요일에 예외 처리 관련해서도 좋은 의견 주셔서 감사했습니다! 고생하셨어요!

99winnmin committed:
refactor: runCatching 범위를 controller -> service layer 로 변경

99winnmin committed:
feat: api end-point logging 을 위한 AOP 설정 및 filter 추가

- Filter 는 request body 추출을 하기 위함, request body 는 InputStream 으로 1회 read 시 휘발되버리므로 ContentCachingRequestWrapper 으로 캐싱할 수 있도록 설정

99winnmin committed:
refactor: api end-point logging 방식 AOP 설정으로 인한 로깅코드 삭제

99winnmin committed:
fix : native query binding 오류로 인한 es 쿼리 수정

99winnmin committed:
fix : code 경고 수정

99winnmin committed:
fix : code 경고 수정
